hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java

```diff
@@ -1195,7 +1196,8 @@ public OzoneOutputStream createKey(
         .setReplicationConfig(replicationConfig)
         .addAllMetadata(metadata)
         .setAcls(getAclList())
-        .setLatestVersionLocation(getLatestVersionLocation);
+        .setLatestVersionLocation(getLatestVersionLocation)
+        .setOwnerName(ownerName);
```

Comment:
Also Missing createStreamFile, createMultipartStremKey for add owner

Reply:
"add owner" has been added, I add this when fixed the ci test

Reply:
Actually I wonder if we really need to add this owner field to every single new key created in Ozone.

I think it might actually better to just leave this as unset (empty) to indicate the default behavior, which is to inherit parent's owner.

Also when a cluster is upgraded to a new Ozone version with key ownership support, all existing keys would natually have empty owners as well (unless we add a lengthy step during OM upgrade to add owner to all existing keys which I don't think is the best idea).

Reply:
@smengcl Thank for your suggestion.
- Necessity for the Owner
"Owner" is a fundamental feature of file systems, including S3 and Filesystem, and is used to explicitly record who created the current Object. This is important for permissions management (Linux UGO and ACL Permission). 
So it is necessary to store a real Owner for each Object. This is the only way to achieve better permissions control and so on.

- For the `chown` command
When we `chown` a directory, we do not recursively `chown` every file in the directory by default, unless the -R option is specified. 

- For the performance
In fact, we ran a performance test after adding an Owner, and there was no observable performance degradation.

- For the owner configuration
Creating a file in Linux, or creating an Object via S3, doesn't have an option to create a file/object with an empty owner, so I think we probably don't need it.

Reply:
Hi @xichen01 , good to know that the perf isn't visibly degraded by this.

1. Back to the `chown` subcommand, the point of the problem I'm raising is that the behavior becomes inconsistent **before and after** an upgrade to an Ozone version with key ownership.

When this key owner addition gets in, and admin `chown` a directory in a bucket, because `chown` is non-recursive by default, the keys inside that dir would still have the old owner field set. This can break some existing Ozone Ranger ACLs that is previously set by admins. e.g. users can be denied access to some keys while in the previous Ozone version they would be allowed acesss, given the same Ranger ACL policy with `{owner}` permission set to ALLOW.

There should be a caveat after Ozone upgrade.

This at least deserves a mention in the **Release Notes**. Please add that to the Apache JIRA accordingly so that the release manager can collect and publish it.

2. Another relevant questions is: How are existing keys WITHOUT an owner field set going to behave after Ozone upgrade to one with key ownership support? Do they inherit parent's owner with the current implementation?

Reply:
Hi @smengcl
> When this key owner addition gets in, and admin chown a directory in a bucket...

The current Ozone key does not have an `owner` field, so the current ACL implementation only checks the `Volume owner` and `Bucket owner`. Adding an `owner` field to the Key will not change this behavior.
So, I don't think it will affect the current ACL, because the current ACL checking process doesn't involve the key owner.

https://github.com/apache/ozone/blob/06c0d81af229d66cbb0464e74dda67b73095d106/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneAclUtils.java#L97-L102

---

> Another relevant questions is:  How are existing keys WITHOUT an owner field set going to behave after Ozone upgrade...

The current `owner` only affects the read APIs such as `listKey`, `getKeyInfo`, etc.  before `owner` in the result of these API were a fake (showing the user who executed the command), now it can show the real `owner`. For old version keys, which didn't have an `owner` field before, it will backoff to the previous logic and populate the `owner` field with the current user executing the command.

Reply:
Thanks @xichen01 for explanation. Got it. So:

1. The current PR hasn't changed the ACL part yet. Thus it doesn't really use the new **key owner** field (yet) during ACL checks.
2. In this PR, the sole purpose of key owner field as of now is to be able to write and read the owner field.

Is my understanding correct so far?

Now I do have a follow-up question:

What is the plan on using the key owner field for ACL checks later on? (Previously I was incorrectly assuming this PR already does what Aswin's bucket owner PR does to ACL checks)

Reply:
> The current PR hasn't changed the ACL part yet. Thus it doesn't really use the new key owner field (yet) during ACL checks.

Right, this PR does not change the ACL check and does not affect the results of the current ACL check because the current ACL check does not even know the key owner existing.

> In this PR, the sole purpose of key owner field as of now is to be able to write and read the owner field.

Right.